### ISSUE TITLE
Operation NOT  permitted in creating /fs/dev/null when using fakeroot on local filesystem

### ISSUE THREAD

**Issue by Amjadhpc** [apptainer/singularity#4826]
### Version of Singularity:
3.4.2
singularity version 3.4.2

### Expected behavior

What did you expect to see when you do...?
singularity build --fakeroot geditfakeroot.sif  gedit.edit

### Actual behavior

What actually happend? Why was it incorrect?

![realsingularityfakeroot](https://user-images.githubusercontent.com/1035011/70154609-b11ce880-16a8-11ea-866b-a4c8db53961b.PNG)

### Steps to reproduce this behavior

How can others reproduce this issue/problem?
 1. Centos make sure /home is not nfs or on GPFS
2. Install singularity 3.4.2 from source
3. make changes in /etc/subuid and  /etc/subgid
4 singularity -d  build --fakeroot geditfakeroot.sif gedit.def.

### What OS/distro are you running
Centos 7
```
$ cat /etc/os-release

CentOS Linux release 7.6.1810 (Core)

```

### How did you install Singularity

From source singularity 3.4.2

**Comment by dtrudg** [comment]
@Amjadhpc  - please check the output of `mount` for your `/tmp` filesystem and let us know what the relevant line is (need to see what type of fs this is, and waht the mount options are).

I suspect that you're using yum here and `/tmp` is mounted `nodev` or similar.

**Comment by Amjadhpc** [comment]
@dctrud 
Here is the output
![image](https://user-images.githubusercontent.com/1035011/70156489-ebd45000-16ab-11ea-893a-9ff3d824c0fd.png)

Yes you are right it is mounted as  nodev. 
P.S  Ignore /run/user/0 and /run/user/64235 those were my attempts to test fakeroot with home dirs on gpfs .

**Comment by dtrudg** [comment]
Okay - the yum bootstrapper cannot be used with a `/tmp` that is mounted `nodev` as bootstrapping a rpm based distribution will create device entries under `/dev`.

You will need to set TMPDIR to a location that is not mounted `nodev`, e.g.

```
TMPDIR=/data/tmpdir singularity build test.sif test.def
```

**Comment by Amjadhpc** [comment]
Wait!!  i think i gave wrong info there . Apologies for it 
/tmp is mounted as xfs  without  nodev option
![image](https://user-images.githubusercontent.com/1035011/70159774-5b007300-16b1-11ea-9b46-e09e8e111448.png)

Or i am missing something here?

**Comment by dtrudg** [comment]
Ahh okay - probably this indicates your singularity version was not build with seccomp support (see https://github.com/sylabs/singularity/issues/3990)

You need to have `libseccomp-devel` installed on RHEL derived systems.

Check for `checking: libseccomp+headers... yes` in the output of `mconfig`

**Comment by Amjadhpc** [comment]
So the question is if libseccomp-devel is not installed , then also singularity >3.4 will build from source without crashing?. Should this not be a mandatory build requirement so that users get fakeroot option  as advertised when building from source ?  I suggest that libssecomp-devel should be a show stopper if it is not installed . Any thoughts on this ?

**Comment by dtrudg** [comment]
We have to support e.g. RHEL6 where there is no user namespace / fakeroot support. It is also possible that installations may not wish to allow seccomp functionality. Bootstrap from `yum` is also not the primary means people are using fakeroot - most will bootstrap from docker or an existing singularity image.

I think we should probably add better information to mconfig, so that it warns e.g. "seccomp support not available - fakeroot yum builds will fail" or similar.

**Comment by Amjadhpc** [comment]
Hi @dctrud 
I can build singularity image using fakeroot  even with my home dir being on /gpfs filesystem.
libpod says (though i have not tested it yet) that  lobpod will not work on home dirs mounted on GPFS. 


https://github.com/containers/libpod/blob/master/rootless.md


[cxe19gxu_s@g02 ~]$ pwd
/gpfs/home/cxe19gxu_s
[cxe19gxu_s@g02 ~]$ singularity --version
singularity version 3.4.2
![image](https://user-images.githubusercontent.com/1035011/70229114-7ae57480-174d-11ea-8a36-39b926b0b3a3.png)

**Comment by dtrudg** [comment]
Glad this works. Building a SIF file is somewhat different to building a sandbox. When you build a SIF file, the initial build of the container filesystem is happening in your TMPDIR, which is not GPFS. The SIF file is then constructed and will be on the GPFS filesystem. Because the SIF file is owned directly by you, and the user-namespace is not involved, it will work on GPFS as you have found.

What you can't do is build a sandbox to GPFS - where the container filesystem is created under the user namespace directly on GPFS. This is the same situation that does not work with libpod.

**Comment by Amjadhpc** [comment]
ok as discussed in slack channel sandboxes  with fakeroot can be build with
version 3.4.2 and not  in 3.5.x . The primary reason was to save time in
building sandboxes on tmpdir and then copying to GPFS . Instead build
directly on GPFS which does not support user namespace.
Not sure what fixes can be seen in 3.5.2 ? .

**Comment by dtrudg** [comment]
This has been opened as an issue here: #4829 

This issue was originally seccomp related and that is being handled here: #4828
